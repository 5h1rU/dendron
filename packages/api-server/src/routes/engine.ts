import {
  DEngineV2,
  EngineQueryRequest,
  EngineWriteRequest,
} from "@dendronhq/common-all";
import { DendronEngine } from "@dendronhq/engine-server";
import { Request, Response, Router } from "express";
import { MemoryStore } from "../store/memoryStore";

const router = Router();

router.get("/health", async (_req: Request, res: Response) => {
  return res.json({ ok: 1 });
});

router.post("/query", async (req: Request, res: Response) => {
  const { ws, queryString, mode } = req.body as EngineQueryRequest;
  const engine = await MemoryStore.instance().get<DendronEngine>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  const { error, data } = await engine.query(queryString, mode);
  const cleanData = data.map((ent) => ent.toRawProps());
  res.json({ error, data: cleanData });
});

router.post("/write", async (req: Request, res: Response) => {
  const { ws, node, opts } = req.body as EngineWriteRequest;
  const engine = await MemoryStore.instance().get<DEngineV2>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  try {
    await engine.writeNote(node, opts);
    res.json({ error: null, data: null });
  } catch (err) {
    res.json({ error: JSON.stringify(err), data: null });
  }
});

export { router as engineRouter };
