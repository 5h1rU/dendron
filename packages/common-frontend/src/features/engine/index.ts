export * as engineReducer from "./slice";
export {
  setError,
  setNotes,
  EngineSliceUtils,
  initNotes,
  InitNoteOpts,
  EngineState,
} from "./slice";
export * as engineStore from "./store";
export {
  RootState as engineState,
  AppDispatch as engineDispatch,
} from "./store";
