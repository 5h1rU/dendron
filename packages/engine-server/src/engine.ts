import {
  DEngine,
  DEngineStore,
  DNodeData,
  DNodeDict,
  DNodeRawProps,
  EngineQueryResp,
  IDNode,
  INoteOpts,
  NodeBuilder,
  NodeWriteOpts,
  Note,
  NoteDict,
  QueryMode,
  QueryOpts,
  Schema,
  SchemaDict,
  SchemaRawProps,
  Scope,
  assert,
  assertExists,
  createLogger,
  getStage,
  makeResponse
} from "@dendronhq/common-all";

import { BodyParser } from "./drivers/raw/BodyParser";
import FileStorage from "./drivers/file/store";
import Fuse from "fuse.js";
import _ from "lodash";
import fs from "fs-extra";

let PROTO_ENGINE: ProtoEngine;
const logger = createLogger("DEngine");

function isAllQuery(qs: string): boolean {
  return qs === "**/*";
}

export interface FuseOptions {
  exactMatch?: boolean;
  preset: QueryMode;
  // isCaseSensitive?: boolean
  // distance?: number
  // findAllMatches?: boolean
  // includeMatches?: boolean
  // includeScore?: boolean
  // location?: number
  // minMatchCharLength?: number
  // shouldSort?: boolean
  // threshold?: number
  // useExtendedSearch?: boolean
}

function createFuse<T>(initList: T[], opts: FuseOptions) {
  const options = {
    shouldSort: true,
    threshold: opts.exactMatch ? 0.0 : 0.6,
    location: 0,
    distance: 100,
    maxPatternLength: 32,
    minMatchCharLength: 1,
    keys: ["title", "path"]
  };
  if (opts.preset === "schema") {
    options.keys = ["title", "id"];
  }
  // initList = _.map(initList, (n) => ({ ...n, treePath: n.path }));
  const fuse = new Fuse(initList, options);
  return fuse;
}

type ProtoEngineOpts = {
  root?: string;
  cacheDir?: string;
  forceNew?: boolean;
};

type ProtoEngineProps = Required<ProtoEngineOpts>;

export class ProtoEngine implements DEngine {
  public fuse: Fuse<Note, any>;
  public schemaFuse: Fuse<Schema, any>;

  public notes: NoteDict;
  public schemas: SchemaDict;

  public fullNodes: Set<string>;

  public opts: ProtoEngineProps;

  public queries: Set<string>;

  public store: DEngineStore;

  static getEngine(opts?: ProtoEngineOpts): DEngine {
    if (!PROTO_ENGINE || opts?.forceNew) {
      const stage = getStage();
      opts = _.defaults(opts || {}, {
        // TODO: remove hardcoded value
        root: "/Users/kevinlin/Dropbox/Apps/Noah/notesv2"
      });
      if (_.isUndefined(opts.root)) {
        throw Error(`root must be defined`);
      }
      // TODO
      PROTO_ENGINE = new ProtoEngine(
        new FileStorage(opts as { root: string }),
        opts
      );
      logger.info({ ctx: "getEngine:exit", opts, stage });
      return PROTO_ENGINE;
    }
    return PROTO_ENGINE;
  }

  constructor(store: DEngineStore, opts: ProtoEngineOpts) {
    // eslint-disable-next-line spaced-comment
    //this.nodes = store.fetchInitial();
    this.notes = {};
    this.store = store;
    this.fuse = createFuse<Note>([], { exactMatch: false, preset: "note" });
    this.fullNodes = new Set();
    this.queries = new Set();

    // setup schemas
    this.schemas = {};
    this.schemaFuse = createFuse<Schema>([], {
      exactMatch: false,
      preset: "schema"
    });

    this.opts = _.defaults(opts, {
      cacheDir: "/tmp/dendronCache",
      root: "/Users/kevinlin/Dropbox/Apps/Dendron",
      forceNew: false
    });
    [opts.cacheDir, opts.root].forEach(fpath => {
      if (!fs.existsSync(fpath as string)) {
        throw Error(`${fpath} doesn't exist`);
      }
    });
  }

  deleteFromNodes(id: string) {
    delete this.notes[id];
    this.fullNodes.delete(id);
  }

  /**
   * Updates local cache
   * - if node is new, add it to cache
   * - if not new, update existing node properties
   * - extra:
   *   - if full node, add id
   * @param nodes
   * @param opts
   */
  refreshNodes(nodes: IDNode[], opts?: QueryOpts) {
    if (_.isEmpty(nodes)) {
      return;
    }
    const mode = nodes[0].type;
    if (mode === "schema") {
      const schemas = nodes as Schema[];
      schemas.forEach((node: Schema) => {
        this.schemas[node.id] = node;
      });
      this.updateLocalCollection(_.values(this.schemas), "schema");
      return;
    } else {
      nodes.forEach((node: IDNode) => {
        const { id } = node;

        if (!_.has(this.notes, id)) {
          // add if not exist
          // TODO: nodes has both raw and full nodes
          // @ts-ignore
          this.notes[id] = node;
        } else {
          // exists, merge it
          logger.debug({
            ctx: "refreshNodes:existingNode",
            node: node.toRawProps()
          });
          _.merge(this.notes[id], node);
        }
        // if a full node, make sure to add
        if (opts?.fullNode) {
          this.fullNodes.add(id);
        }
      });
      // FIXME: debug
      const newNodes = _.map(nodes, n => ({ title: n.title, id: n.id }));
      const allNodes = _.map(this.notes, n => ({
        title: n.title,
        id: n.id
      }));
      logger.debug({
        ctx: "refreshNodes",
        newNodes,
        allNodes
      });
      this.updateLocalCollection(_.values(this.notes), "note");
      return;
    }
  }

  /**
   * Turn parent|children ids to full nodes
   */
  async resolveIds(
    node: DNodeRawProps<DNodeData>,
    dict: DNodeDict
  ): Promise<Note> {
    const nodeResolved: INoteOpts = dict[node.id] as INoteOpts;
    assert(
      !_.isUndefined(nodeResolved),
      `node ${JSON.stringify(node)} not in dict\nkeys: ${_.keys(dict)}`
    );
    nodeResolved.body = node.body;
    return new Note({ ...nodeResolved });
  }

  updateLocalCollection(collection: IDNode[], mode: QueryMode) {
    if (mode === "schema") {
      return this.schemaFuse.setCollection(collection as Schema[]);
    } else {
      return this.fuse.setCollection(collection as Note[]);
    }
  }

  async delete(id: string): Promise<void> {
    await this.store.delete(id);
    this.deleteFromNodes(id);
    return;
  }

  /**
   * Defaults
   * @param id
   * @param opts
   *  - fullNode: {default:true}
   */
  async get(_scope: Scope, id: string, mode: QueryMode, opts?: QueryOpts) {
    opts = _.defaults(opts || {}, { fullNode: true, createIfNew: true });
    let nodeDict;
    logger.debug({ ctx: "get", id, opts });

    if (mode === "schema") {
      nodeDict = this.schemas;
      const node = nodeDict[id];
      return { data: node };
    }
    //FIXME: check if exist
    // we are assuming all nodes are fetched
    const node = this.notes[id];

    // a full node has a body and is fully resolved
    if (opts?.fullNode && !this.fullNodes.has(id)) {
      logger.debug({ ctx: "get:fetchFromStore:pre", id });
      const fnResp = await this.store.get(_scope, id, {
        ...opts,
        webClient: true
      });
      logger.debug({ ctx: "get:fetchFromStore:post", id, opts, fnResp });
      const fullNode = await this.resolveIds(fnResp.data, this.notes);
      logger.debug({ ctx: "get:resolve:post", fnResp });
      // TODO:
      this.refreshNodes([fullNode], opts);
      return { data: fullNode };
    }
    logger.debug({ ctx: "get:exit", node });
    return { data: node };
  }

  async query(
    scope: Scope,
    queryString: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) {
    opts = _.defaults(opts || {}, {
      fullNode: false,
      createIfNew: true,
      initialQuery: false
    });
    let data: EngineQueryResp;

    // handle all query case
    if (isAllQuery(queryString)) {
      logger.debug({ ctx: "query:queryAll:pre", mode });
      data = await this.store.query(scope, "**/*", mode, opts);
      if (opts.initialQuery) {
        this.refreshNodes(data.data);
      }
      // @ts-ignore
      return data as EngineQueryResp;
    }

    // handle schema query
    if (mode === "schema") {
      const results = this.schemaFuse.search(queryString);
      let items: Schema[];
      const nodes = this.schemas;
      if (opts.queryOne) {
        items = [results[0]?.item];
      } else {
        items = _.map(results, resp => resp.item);
      }
      logger.debug({ ctx: "query:exit:schema", items });
      return makeResponse<EngineQueryResp>({
        data: _.map(items, item => nodes[item.id]),
        error: null
      });
    } else {
      // handle note query
      const results = this.fuse.search(queryString);
      let items: IDNode[];
      if (opts.queryOne) {
        items = [results[0]?.item];
      } else {
        items = _.map(results, resp => resp.item);
      }

      if (opts.queryOne && items[0]?.path !== queryString && opts.createIfNew) {
        logger.debug({
          ctx: "query:write:pre",
          queryString,
          item: items[0]
        });
        const nodeBlank = new Note({ fname: queryString });
        await this.write(scope, nodeBlank, { newNode: true });
        const { data: nodeFull } = await this.get(scope, nodeBlank.id, mode);
        this.refreshNodes([nodeFull], { fullNode: true });
        return makeResponse<EngineQueryResp>({
          data: [nodeFull],
          error: null
        });
      }

      if (opts.fullNode) {
        const fetchedFullNodes = await Promise.all(
          _.map<IDNode, Promise<IDNode | null>>(items, async ent => {
            if (!this.fullNodes.has(ent.id)) {
              logger.debug({
                ctx: "query:fuse.search:post",
                status: "fetch full node from store"
              });
              // FIXME: ratelimit
              const fn = await this.get(scope, ent.id, mode);
              return fn.data;
            }
            logger.debug({
              ctx: "query:fuse.search:post",
              status: "fetch full node from cache"
            });
            return null;
          })
        );
        this.refreshNodes(
          _.filter(fetchedFullNodes, ent => !_.isNull(ent)) as IDNode[],
          { fullNode: true }
        );
        logger.debug({
          ctx: "query:fetchedFullNodes:exit",
          fetchedFullNodes
        });
      }
      logger.debug({ ctx: "query:exit:note", items });
      return makeResponse<EngineQueryResp>({
        data: _.map(items, item => this.notes[item.id]),
        error: null
      });
    }
  }

  async write(scope: Scope, node: IDNode, opts?: NodeWriteOpts): Promise<void> {
    opts = _.defaults(opts, { newNode: false, body: "" });
    if (node.type === "schema") {
      assertExists(opts.body, "body must exist");
      // convert body
      const props: SchemaRawProps[] = new BodyParser().parseSchema(
        opts.body as string,
        {
          node,
          fname: node.fname
        }
      );
      const schemas = new NodeBuilder().buildSchemaFromProps(props);
      const schemaNew = _.find(schemas, { id: node.id });
      if (_.isUndefined(schemaNew)) {
        throw Error(`no schema found for ${node} on write`);
      }
      node = schemaNew;
      // reset body
      node.body = "";
      await this.store.write(scope, node);
      if (opts.newNode) {
        const parentPath = "root";
        const parentNode = _.find(this.schemas, n => n.title === parentPath);
        if (!parentNode) {
          throw Error("no parent found");
        }
        (parentNode as Schema).addChild(node as Schema);
      }
    } else {
      const note = node as Note;
      await this.store.write(scope, note);
      if (opts.newNode) {
        let parentPath = note.fname
          .split(".")
          .slice(0, -1)
          .join(".");
        if (_.isEmpty(parentPath)) {
          parentPath = "root";
        }
        const parentNode = _.find(this.notes, n => n.path === parentPath);
        if (!parentNode) {
          throw Error("no parent found");
        }
        parentNode.addChild(note);
      }
    }
    // TODO
    return this.refreshNodes([node], { fullNode: opts.newNode });
  }
}

export function engine() {
  return ProtoEngine.getEngine();
}
