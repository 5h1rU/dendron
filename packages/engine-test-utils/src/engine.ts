import { DEngineClientV2, DVault, WorkspaceOpts } from "@dendronhq/common-all";
import { launch } from "@dendronhq/api-server";
import { tmpDir } from "@dendronhq/common-server";
import {
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
} from "@dendronhq/common-test-utils";
import {
  createEngine as engineServerCreateEngine,
  DendronEngineClient,
  WorkspaceService,
} from "@dendronhq/engine-server";
import _ from "lodash";

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<DEngineClientV2>;

export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return engineServerCreateEngine(opts);
}

export { DEngineClientV2, DVault, WorkspaceOpts };

export async function createEngineFromServer(opts: WorkspaceOpts) {
  const port = await launch({});
  const engine: DEngineClientV2 = DendronEngineClient.create({
    port,
    vaultsv4: opts.vaults,
    ws: opts.wsRoot,
    vaults: [],
  });
  await engine.init();
  return engine;
}

async function setupWS(opts: { vaults: DVault[] }) {
  const wsRoot = tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  const vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault });
      return vault;
    })
  );
  return { wsRoot, vaults };
}

// @ts-ignore
export async function runEngineTestV5<TExtra = any>(
  func: RunEngineTestFunctionV4,
  opts: {
    preSetupHook?: SetupHookFunction;
    //postSetupHook?: PostSetupHookFunction;
    createEngine?: AsyncCreateEngineFunction;
    extra?: TExtra;
    expect: any;
    vaults?: DVault[];
    setupOnly?: boolean;
  }
): Promise<any> {
  const { preSetupHook, extra, vaults, createEngine } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    vaults: [{ fsPath: "vault1" }, { fsPath: "vault2" }],
  });
  const { wsRoot } = await setupWS({ vaults });
  await preSetupHook({ wsRoot, vaults });
  const engine = await createEngine({ wsRoot, vaults });
  const initResp = await engine.init();
  const testOpts = { wsRoot, vaults, engine, initResp, extra };
  if (opts.setupOnly) {
    return testOpts;
  }
  const results = (await func(testOpts)) || [];
  await runJestHarnessV2(results, expect);
  return { opts: testOpts, resp: undefined };
}
