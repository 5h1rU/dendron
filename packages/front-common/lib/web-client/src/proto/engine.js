var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { DropboxStorage, Note, makeResponse } from "@dendron/common-all";
import Fuse from "fuse.js";
import { Logger } from "@aws-amplify/core";
import _ from "lodash";
var logger = new Logger("DEngine");
var PROTO_ENGINE;
function createMockData() {
    var _a;
    var secondChildNote = new Note({
        id: "manifesto",
        title: "manifesto",
        desc: "first child desc",
        type: "note",
        schemaId: "-1",
    });
    var firstChildNote = new Note({
        id: "dendron",
        title: "dendron",
        desc: "first child desc",
        type: "note",
        schemaId: "-1",
    });
    var rootNote = new Note({
        id: "root",
        title: "root",
        desc: "root",
        type: "note",
        schemaId: "-1",
    });
    rootNote.addChild(firstChildNote);
    firstChildNote.addChild(secondChildNote);
    var initialNodes = (_a = {},
        _a[rootNote.id] = rootNote,
        _a[firstChildNote.id] = firstChildNote,
        _a[secondChildNote.id] = secondChildNote,
        _a);
    return initialNodes;
}
var INITIAL_DATA = createMockData();
function createFuse(initList, opts) {
    var options = {
        shouldSort: true,
        threshold: opts.exactMatch ? 0.0 : 0.6,
        location: 0,
        distance: 100,
        maxPatternLength: 32,
        minMatchCharLength: 1,
        keys: ["title", "path"],
    };
    // initList = _.map(initList, (n) => ({ ...n, treePath: n.path }));
    // console.log({ initList, bond: true });
    var fuse = new Fuse(initList, options);
    return fuse;
}
// eslint-disable-next-line @typescript-eslint/no-unused-vars
// @ts-ignore
var MockDataStore = /** @class */ (function () {
    function MockDataStore() {
        this.data = INITIAL_DATA;
    }
    MockDataStore.prototype.fetchInitial = function () {
        return this.data;
    };
    MockDataStore.prototype.get = function (_scope, id) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                // TODO
                return [2 /*return*/, new Promise(function (resolve) {
                        setTimeout(function () {
                            var note = _.get(INITIAL_DATA, id, new Note({
                                id: id,
                                title: "sample",
                                desc: "sample",
                                type: "note",
                                schemaId: "-1",
                            }));
                            note.body = "content for " + id;
                            resolve({ data: note });
                        }, 1200);
                    })];
            });
        });
    };
    return MockDataStore;
}());
export { MockDataStore };
// @ts-ignore - TODO: implement interface
var ProtoEngine = /** @class */ (function () {
    function ProtoEngine(store) {
        //this.nodes = store.fetchInitial();
        this.nodes = {};
        this.store = store;
        var fuseList = _.values(this.nodes);
        this.fuse = createFuse(fuseList, { exactMatch: false });
        this.fullNodes = new Set();
        this.queries = new Set();
    }
    ProtoEngine.getEngine = function () {
        if (!PROTO_ENGINE) {
            // PROTO_ENGINE = new ProtoEngine(new MockDataStore());
            PROTO_ENGINE = new ProtoEngine(new DropboxStorage());
            return PROTO_ENGINE;
        }
        return PROTO_ENGINE;
    };
    ProtoEngine.prototype._nodeInCache = function (node, opts) {
        var _a;
        var hasStub = _.has(this.nodes, node.id);
        var fufillsFull = ((_a = opts) === null || _a === void 0 ? void 0 : _a.fullNode) ? true : this.fullNodes.has(node.id);
        return hasStub && fufillsFull;
    };
    // FIXME: query doesn't check for full nodes
    ProtoEngine.prototype._queryInCache = function (qs) {
        var hasQuery = _.has(this.queries, qs);
        return hasQuery;
    };
    /**
     * Updates local cache
     * @param nodes
     * @param opts
     */
    ProtoEngine.prototype.refreshNodes = function (nodes, opts) {
        var _this = this;
        nodes.forEach(function (node) {
            var _a;
            var id = node.id;
            // add if not exist
            if (!_.has(_this.nodes, id)) {
                _this.nodes[id] = node;
            }
            else {
                _.merge(_this.nodes[id], node);
                if ((_a = opts) === null || _a === void 0 ? void 0 : _a.fullNode) {
                    _this.fullNodes.add(id);
                }
            }
        });
        this.updateLocalCollection(_.values(this.nodes));
    };
    ProtoEngine.prototype.updateLocalCollection = function (collection) {
        this.fuse.setCollection(collection);
    };
    ProtoEngine.prototype.get = function (_scope, id, opts) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var node, fnResp;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        node = this.nodes[id];
                        opts = _.defaults(opts || {}, { fullNode: true });
                        if (!(((_a = opts) === null || _a === void 0 ? void 0 : _a.fullNode) && !this.fullNodes.has(id))) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.store.get(_scope, id, __assign(__assign({}, opts), { webClient: true }))];
                    case 1:
                        fnResp = _b.sent();
                        logger.debug({ ctx: "get:store.get:post", id: id, opts: opts, fnResp: fnResp });
                        this.refreshNodes([fnResp.data], opts);
                        return [2 /*return*/, fnResp];
                    case 2: return [2 /*return*/, { data: node }];
                }
            });
        });
    };
    ProtoEngine.prototype.query = function (scope, queryString, opts) {
        return __awaiter(this, void 0, void 0, function () {
            var data, results, items, fetchedFullNodes;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        console.log({ scope: scope, queryString: queryString, opts: opts });
                        opts = _.defaults(opts || {}, {
                            fullNode: false,
                        });
                        if (!(queryString === "**/*")) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.store.query(scope, "**/*", opts)];
                    case 1:
                        data = _a.sent();
                        this.refreshNodes(data.data);
                        return [2 /*return*/, data];
                    case 2:
                        results = this.fuse.search(queryString);
                        logger.debug({ ctx: "query:fuse.search:post", results: results, queryString: queryString });
                        if (opts.queryOne) {
                            items = [results[0].item];
                        }
                        else {
                            items = _.map(results, function (resp) { return resp.item; });
                        }
                        if (!opts.fullNode) return [3 /*break*/, 4];
                        return [4 /*yield*/, Promise.all(_.map(items, function (ent) { return __awaiter(_this, void 0, void 0, function () {
                                var fn;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            if (!!this.fullNodes.has(ent.id)) return [3 /*break*/, 2];
                                            logger.debug({
                                                ctx: "query:fuse.search:post",
                                                status: "fetch full node from store",
                                            });
                                            return [4 /*yield*/, this.get(scope, ent.id)];
                                        case 1:
                                            fn = _a.sent();
                                            return [2 /*return*/, fn.data];
                                        case 2:
                                            logger.debug({
                                                ctx: "query:fuse.search:post",
                                                status: "fetch full node from cache",
                                            });
                                            return [2 /*return*/, null];
                                    }
                                });
                            }); }))];
                    case 3:
                        fetchedFullNodes = _a.sent();
                        this.refreshNodes(_.filter(fetchedFullNodes, function (ent) { return !_.isNull(ent); }), { fullNode: true });
                        logger.debug({ ctx: "query:fetchedFullNodes:exit", fetchedFullNodes: fetchedFullNodes });
                        _a.label = 4;
                    case 4:
                        logger.debug({ ctx: "query:exit", items: items });
                        return [2 /*return*/, makeResponse({
                                data: _.map(items, function (item) { return _this.nodes[item.id]; }),
                                error: null,
                            })];
                }
            });
        });
    };
    return ProtoEngine;
}());
export { ProtoEngine };
export function engine() {
    return ProtoEngine.getEngine();
}
//# sourceMappingURL=engine.js.map