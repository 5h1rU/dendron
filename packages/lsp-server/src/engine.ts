import { DendronEngine } from "@dendronhq/engine-server";
export { DendronEngine };
