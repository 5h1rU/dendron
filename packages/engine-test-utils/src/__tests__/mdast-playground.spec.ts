import {
  DendronASTDest,
  mdastBuilder,
  MDUtilsV4,
} from "@dendronhq/engine-server";
import { testWithEngine } from "../engine";
const { root, paragraph, link, listItem, text, heading, list } = mdastBuilder;

describe("proto", () => {
  test("basic", () => {
    const proc = MDUtilsV4.remark();

    const output = proc.stringify(
      root([
        heading(2, text("Begin")),
        paragraph([
          list("unordered", [
            paragraph(link("dendron.so", undefined, text("bullet link"))),
            listItem(text("one")),
            listItem(text("two")),
            listItem(text("three")),
          ]),
        ]),
      ])
    );
    expect(output).toMatchSnapshot();
  });

  test("basic2", async () => {
    const proc = MDUtilsV4.remark();

    const output = await proc.parse(["line one", "***", "line two"].join("\n"));
    expect(output).toMatchSnapshot();
  });

  testWithEngine("wiilinks as list", async ({ engine, vaults }) => {
    const proc = MDUtilsV4.procFull({
      dest: DendronASTDest.MD_DENDRON,
      engine,
      fname: "foo",
      vault: vaults[0],
    });
    const output = await proc.parse(["- [[one]]", "- [[two]]"].join("\n"));
    expect(output).toMatchSnapshot();
  });
});
