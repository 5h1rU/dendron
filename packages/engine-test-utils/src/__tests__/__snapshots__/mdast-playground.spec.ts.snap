// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proto basic 1`] = `
"## Begin

- [bullet link](dendron.so)
- one
- two
- three
"
`;

exports[`proto basic2 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "line one",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 9,
          "line": 1,
          "offset": 8,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 2,
          "offset": 12,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 9,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 9,
              "line": 3,
              "offset": 21,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 13,
            },
          },
          "type": "text",
          "value": "line two",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 9,
          "line": 3,
          "offset": 21,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 13,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 9,
      "line": 3,
      "offset": 21,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`proto wiilinks as list 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "alias": "one",
                    "anchorHeader": undefined,
                    "sameFile": false,
                    "vaultName": undefined,
                  },
                  "position": Position {
                    "end": Object {
                      "column": 10,
                      "line": 1,
                      "offset": 9,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 1,
                      "offset": 2,
                    },
                  },
                  "type": "wikiLink",
                  "value": "one",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "alias": "two",
                    "anchorHeader": undefined,
                    "sameFile": false,
                    "vaultName": undefined,
                  },
                  "position": Position {
                    "end": Object {
                      "column": 10,
                      "line": 2,
                      "offset": 19,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 2,
                      "offset": 12,
                    },
                  },
                  "type": "wikiLink",
                  "value": "two",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 10,
                  "line": 2,
                  "offset": 19,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 2,
                  "offset": 12,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 10,
              "line": 2,
              "offset": 19,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 10,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": Position {
        "end": Object {
          "column": 10,
          "line": 2,
          "offset": 19,
        },
        "indent": Array [
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 10,
      "line": 2,
      "offset": 19,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
